### ISSUE TITLE
Profile  LDS_MaintainSimplifiedLayers and LDS_MaintainAllFBDELayers

### ISSUE THREAD

**Issue by strk** [linz/linz-lds-bde-schema#20]
According to a [profiling run](https://github.com/linz/bde-processor-deployment/issues/28#issuecomment-315203412) It is possible that 2GB are not enough to run LDS_MaintainSimplifiedLayers.

This ticket is to review the function to spot any unoptimized memory usage.

**Comment by strk** [comment]
I've set Vagrant machine memory to 1GB and had run the function. The function takes an upload identifier and based on that does more or less things, based on whether the identified upload changed or not some tables. To my surprise, it took 30 seconds to run when passing the identifier of an upload which has NO records in `bde_control.upload_table`. The function run logs many occurrences of `0 rows were created`, but still it takes 30 seconds to run. Is this expected @palmerj  ?

**Comment by palmerj** [comment]
> Is this expected @palmerj ?

No it's not! Each of the sub functions should be checking that tables have changed state for the given upload_id via bde_control.bde_TablesAffected before proceeding to anything that actually does processing. Is the upload_id isn't real then it. Are you just calling the function something like this:

```plpgsql
SELECT LDS_MaintainSimplifiedLayers(-1);
```

?

Maybe turn on profiling for functions just for the LDS_MaintainSimplifiedLayers run and see what functions get called and for what time.

**Comment by strk** [comment]
`-1` is not accepted (identifiers aren't quoted).
I'm using `SELECT lds.LDS_MaintainSimplifiedLayers(1)`.
I see it actually creates 12 rows, bu tit takes 12.5 seconds to do so:
```
vagrant@bde-processor:~$ time psql -c 'SELECT lds.LDS_MaintainSimplifiedLayers(1)' linz 2>&1 | grep 'rows were created'  | grep -v '0 rows'
INFO:  Finished creating new version of table land_districts for differencing. 12 rows were created

real    0m12.584s
user    0m0.038s
sys     0m0.026s
```

I saw CREATE INDEX notices, when not filtering out output.

These are the stats for that single call:
```
linz=# select * from pg_stat_user_functions order by total_time desc limit 10;
 funcid |  schemaname   |               funcname               | calls | total_time | self_time 
--------+---------------+--------------------------------------+-------+------------+-----------
  19820 | lds           | lds_maintainsimplifiedlayers         |     1 |   12752.08 |     0.681
  19837 | lds           | lds_maintainsimplifiedgeodeticlayers |     1 |  12359.244 | 12277.887
  19822 | lds           | lds_updatesimplifiedtable            |    44 |    384.643 |    83.834
  19847 | lds           | lds_maintainsimplifiedsurveylayers   |     1 |    254.506 |    12.726
  53733 | bde_control   | bde_executetemplate                  |    44 |    163.284 |   162.364
  19840 | lds           | lds_maintainsimplifiedparcellayers   |     1 |    109.631 |     25.14
  53740 | bde_control   | bde_executesqlarray                  |    43 |     66.694 |    66.694
  19829 | lds           | lds_applytabledifferences            |     1 |     46.454 |     0.195
  18147 | table_version | ver_apply_table_differences          |     1 |     46.259 |     1.164
  18133 | table_version | ver_executetemplate                  |     1 |     44.745 |     0.493
(10 rows)
```

**Comment by strk** [comment]
I haven't had memory issues with this function recently so I suspect when I had them it was related to having less memory available than configured. I'm turning this ticket into a generic profiling task for the still most expensive functions of BDE processor.

**Comment by strk** [comment]
Profile info for `LDS_MaintainSimplifiedSurveyLayers`, which is one of the functions called by `LDS_MaintainSimplifiedLayers` is here: https://github.com/linz/bde-processor-deployment/wiki/MaintainSimplifiedSurveyLayers-Profiling-2017

**Comment by strk** [comment]
Entry point for LDS_SimplifiedLayers profiling: https://github.com/linz/bde-processor-deployment/wiki/MaintainSimplifiedLayers-Profiling-2017

**Comment by strk** [comment]
Entry point for LDS_MaintainAllFBDELayers profiling: https://github.com/linz/bde-processor-deployment/wiki/MaintainAllFBDELayers-Profiling-2017

**Comment by strk** [comment]
I've updated the wiki pages to also show numbers for the most expensive child of `MaintainSimplifiedLayers`:  https://github.com/linz/bde-processor-deployment/wiki/MaintainSimplifiedParcelLayers-Profiling-2017

That function takes 1 hour to run, same as the `MaintainAllFBDELayers` one.

Then I went chasing the most expensive part of parcel layers maintainance, which is preparation of the temporary parcel table, and found a jungle of temporary tables created in order to get to the final `tmp_parcel` table. One of these temporary tables costed 15 seconds and I found that it could run in half a second i fit wasn't for a plpgsql function called for each and every record: `bde.bde_get_par_stat_act()`. There are many such cases of `plpgsql` functions called for every record of a table, enough that there are (as the wiki page shows) over 11 million calls to `bde.bde_get_combined_appellation`, for example. Those calls alone represent the 28% of all the time spent by `MaintainSimplifiedParcelLayers`. Given those functions run further selects on tables, I wonder if they could be inlined to give the planner a way to reduce the number of table scans, but I hadn't gone so deep in understanding each of those functions. Adding [documentation](https://github.com/linz/linz-bde-schema/issues/26 ) to them could help.
